lisp/ess-r-mode.el

```diff
@@ -2352,7 +2352,8 @@ If the current buffer does not have a usage section, return nil."
           (goto-char (point-min))
           (forward-whitespace 1)
           (while (not (eobp))
-            (if (looking-at (rx (group (1+ (not (any whitespace)))) "("))
+            (if (looking-at "#") (forward-line 1))
+            (if (looking-at (rx (group (1+ (not (any "#" whitespace)))) "("))
```

Comment:
This seems a bit brittle compared to the approach of parsing on the R side. But we could merge this as a stopgap, it's probably good enough for most cases. I'm a bit hesitant making more changes without unit tests. It seems it would not be too difficult to include some cases in `test/fixtures` files and apply that routine to buffers visiting those files?

Reply:
Brittle: I see it as simple, not fragile, I think I'm missing something. I've tested this elisp on several help-pages (see my next comment), and I could not think of or find a casualty. Do you have ideas (or examples) of help docs (or classes/types/structures) that would break with this?

If you mean something more meaningful than just "comments", the problem-case-du-jour starts with `# DT[J(...)]`, which was previously parsed as `DT[J`. Knowing what `data.table` *means* by this, I don't know how one could infer their intent from this. The fact that it is commented-out leaves us one escape-hatch; the only other alternative I can see (if it were not commented-out) would be to use `"try(names(formals('" u "')))\n"` ... which I still think is a worthwhile addition, but I defer to your judgement.

Tests: I understand, though I am woefully unprepared for elisp unit-testing. I imagine that the top portion of a few help-docs can work for `test/fixtures`, perhaps including `data.table::J` (or fake samples structured similarly). But beyond that, I don't know if I can construct elisp-tests without a lot of hand-holding. Sorry, I may be more of a liability for this.

Parsing on the R side: I think you're suggesting a new R function that, given the help text (whole or just the "Usage" section), returns the list of functions. This would replace most of the middle of `ess-r-help-usage-objects`, picking up again with `(when usage-objects ...)`. Something like:

```elisp
(defun ess-r-help-usage-objects ()
  "Return a list of objects in the usage section for the current help buffer.
In other words, if in the help buffer for \"qt\", return

'((\"dt\" \"x\" \"df\" \"ncp\" \"log\")
  (\"pt\" \"q\" \"df\" \"ncp\" \"lower.tail\" \"log.p\")
  (\"qt\" \"p\" \"df\" \"ncp\" \"lower.tail\" \"log.p\")
  (\"rt\" \"n\" \"df\" \"ncp\")).

If the current buffer does not have a usage section, return nil."
  (unless (derived-mode-p 'ess-r-help-mode)
    (error "Not an R help buffer"))
  (save-excursion
    (save-restriction
      (when-let ((usage-objects (ess-get-words-from-vector (concat ".ess_somenewfunction(' " (buffer-string) " ')\n"))))
        (when usage-objects
          ;; Get arguments:
          (setq usage-objects
                (mapcar (lambda (u) (cons u (ess-get-words-from-vector (concat "names(formals(" u "))\n"))))
                        usage-objects)))
        (nreverse usage-objects)))))
```

I think I need to quote-escape `(buffer-string)`, but does this look about right?

Reply:
Regarding robustness, we might get in trouble with non-syntactic names for instance, or with more complex code. I'm not sure what's allowed in these usage sections. But no worries, it's probably good enough for the vast majority of cases.

Regarding unit tests, if you can prepare a few good cases, I could pick it up from there. Just a case with a normal usage section, and then a few cases involving comments.

Out of curiosity, do you get lots of failures when you run `make` inside the `test` folder? We have never run it on Windows.

Reply:
Okay, sticking with elisp for now, no new R functions.

Unit tests: thanks. Do you have issues with me including the partial samples I included in the other comment? I'll just add them to the commit. (Notably, `read.table` is multi-line no comments, and either of the `J` samples I gave have comments of various evil-ness. We can always include "data" and/or "package" non-function help pages if you think we need to go that far. They both seem to parse just fine with the `cond` code.)

`make`: For perhaps "obvious" reasons, I tend to `make` very infrequently in windows, preferring instead the linux servers I do "real" work on (remotely over ssh, less interactive use).

Side note, tests work much better when I remove incomplete scratch code from `ess-r-mode.el` ... ;-)

In WSL2 (running ubuntu), I get two fails (and five skips):

```
Test ess-test-r-comint-input-ring-file-name backtrace:
  insert-file-contents("/tmp/foo")
  ...
Test ess-test-r-comint-input-ring-file-name condition:
    (file-error "Read error" "Is a directory" "/tmp/foo")
   FAILED   85/125  ess-test-r-comint-input-ring-file-name
```

I have a `foo` directory within `/tmp`, for whatever reason. Moving that directory out of the way fixes this error. This suggests that some test is "unsafe", though I can't find it.

The other fail is:

```
Test ess-test-r-help-mode condition:
    (ert-test-failed
     ((should
       (or
        (equal ess-help-object "plot.default")
        (equal ess-help-object "plot")))
      :form
      (or
       (equal ess-help-object "plot.default")
       (equal ess-help-object "plot"))
      :value nil))
   FAILED   86/125  ess-test-r-help-mode
```

This also fails in the `master` branch, so seems unlikely due to these edits.

---

If you're curious, windows itself (not WSL2) did not fare so well.

```
9 unexpected results:
   FAILED  ess--command-browser-timeout-test
   FAILED  ess-r-eval-visibility-eval-test
   FAILED  ess-r-package-vars-test
   FAILED  ess-r-pager-test
   FAILED  ess-set-working-directory-test
   FAILED  ess-startup-verbose-setwd-test
   FAILED  ess-test-inferior-local-start-args
   FAILED  ess-test-inferior-reload-start-data
   FAILED  ess-test-r-help-mode
```

That's not a rabbit-hole I want to go down right now. :-)

<details><summary>In case you are curious/bored ...</summary>

```
make[1]: Entering directory '/Users/r2/Projects/github/ESS/_1025'
********************* VERSIONS **************************
GNU Emacs 26.3
ESS 18.10.3snapshot
ESSR 1.6
*********************************************************
make -C lisp all
make[2]: Entering directory '/Users/r2/Projects/github/ESS/_1025/lisp'
make[2]: Nothing to be done for 'all'.
make[2]: Leaving directory '/Users/r2/Projects/github/ESS/_1025/lisp'
make[1]: Leaving directory '/Users/r2/Projects/github/ESS/_1025'
make[1]: Entering directory '/Users/r2/Projects/github/ESS/_1025'
make[2]: Entering directory '/Users/r2/Projects/github/ESS/_1025/lisp'
make[2]: 'julia-mode.el' is up to date.
make[2]: Leaving directory '/Users/r2/Projects/github/ESS/_1025/lisp'
make[1]: Leaving directory '/Users/r2/Projects/github/ESS/_1025'
emacs -Q --script run-tests
Running 125 tests (2020-12-30 11:44:21-0500)
   passed    1/125  R-mode
Test ess--command-browser-timeout-test backtrace:
  signal(ert-test-failed (((should (string= actual expected)) :form (string= "Called from: top level \nBrowse[1]> " "Called from: 
  ert-fail(((should (string= actual expected)) :form (string= "Called from: top level \nBrowse[1]> " "Called from: top level \nBro
  (if (unwind-protect (setq value-125 (apply fn-123 args-124)) (setq form-description-127 (nconc (list (quote (should (string= act
  (let (form-description-127) (if (unwind-protect (setq value-125 (apply fn-123 args-124)) (setq form-description-127 (nconc (list
  (let ((value-125 (quote ert-form-evaluation-aborted-126))) (let (form-description-127) (if (unwind-protect (setq value-125 (appl
  (let* ((fn-123 (function string=)) (args-124 (condition-case err (let ((signal-hook-function (function ert--should-signal-hook))
  (closure (t) (actual expected) (let* ((fn-123 (function string=)) (args-124 (condition-case err (let ((signal-hook-function ...)
  funcall((closure (t) (actual expected) (let* ((fn-123 (function string=)) (args-124 (condition-case err (let (...) (list actual 
  (save-current-buffer (set-buffer inf-buf) (save-excursion (let ((times 0)) (while (and (< times 10) (not (re-search-forward "> \
  (let* ((inf-buf etest-local-inferior-buffer) (inf-proc (get-buffer-process inf-buf))) (save-current-buffer (set-buffer inf-buf) 
  (unwind-protect (let* ((inf-buf etest-local-inferior-buffer) (inf-proc (get-buffer-process inf-buf))) (save-current-buffer (set-
  etest--flush-inferior-buffer((closure (t) (actual expected) (let* ((fn-123 (function string=)) (args-124 (condition-case err (le
  (cond ((eq etest--key :inf-buffer) (setq etest-local-inferior-buffer (eval etest--value))) ((eq etest--key :cleanup) (setq etest
  (let ((etest--key (car-safe (prog1 body (setq body (cdr body))))) (etest--value (car-safe (prog1 body (setq body (cdr body))))))
  (while body (let ((etest--key (car-safe (prog1 body (setq body (cdr body))))) (etest--value (car-safe (prog1 body (setq body (cd
  (unwind-protect (while body (let ((etest--key (car-safe (prog1 body (setq body ...)))) (etest--value (car-safe (prog1 body (setq
  (let ((etest--msg-sentinel (etest--make-message-sentinel)) etest--cleanup) (unwind-protect (while body (let ((etest--key (car-sa
  (save-current-buffer (set-buffer etest--buf) (let ((etest--msg-sentinel (etest--make-message-sentinel)) etest--cleanup) (unwind-
  (unwind-protect (save-current-buffer (set-buffer etest--buf) (let ((etest--msg-sentinel (etest--make-message-sentinel)) etest--c
  (let ((etest--buf (etest--new-buffer (let (local) (while (eq (car body) :init) (car-safe (prog1 body ...)) (setq local (append l
  etest--run-test((:init ((mode . r) (eval ess-test-r-set-local-process)) :cleanup (progn (ess-debug-command-quit) (ess-wait-for-p
  (closure (t) nil (etest--run-test (quote (:init ((mode . r) (eval ess-test-r-set-local-process)) :cleanup (progn (ess-debug-comm
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test :name ess--command-browser-timeout-test :documentation "`es
  ert-run-test(#s(ert-test :name ess--command-browser-timeout-test :documentation "`ess-command' fails with hanging command within
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test :name R-mode :documentation nil :body (closure (t) nil (let*
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #<bytecode 0x10028b491>) nil)
  ert-run-tests-batch(t)
  ert-run-tests-batch-and-exit(t)
  eval-buffer(#<buffer  *load*> nil "c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t)  ; Reading at buffer position 23
  load-with-code-conversion("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" "c:/Users/r2/Projects/github/ESS/_1025/test/run
  load("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t t)
  command-line-1(("-scriptload" "run-tests"))
  command-line()
  normal-top-level()
Test ess--command-browser-timeout-test condition:
    (ert-test-failed
     ((should
       (string= actual expected))
      :form
      (string= "Called from: top level 
Browse[1]> " "Called from: top level 
Browse[1]> debug at #1: NULL
Browse[1]> ")
      :value nil))
   FAILED    2/125  ess--command-browser-timeout-test
   passed    3/125  ess--derive-connection-path
   passed    4/125  ess-Rout-file-test
   passed    5/125  ess-async-command-test
   passed    6/125  ess-bugs-mode
   passed    7/125  ess-build-eval-command-R-test
   passed    8/125  ess-build-eval-command-test
   passed    9/125  ess-build-load-command-R-test
   passed   10/125  ess-build-load-command-test
   passed   11/125  ess-command-hanging-test
   passed   12/125  ess-command-incomplete-test
   passed   13/125  ess-command-last-value-test
   passed   14/125  ess-command-test
   passed   15/125  ess-cycle-assign-test
   passed   16/125  ess-eval-and-step-test
   passed   17/125  ess-eval-function
   passed   18/125  ess-eval-line-test
   passed   19/125  ess-eval-paragraph-test
   passed   20/125  ess-eval-region-test
   passed   21/125  ess-find-newest-date-test
   passed   22/125  ess-flatten-list
   passed   23/125  ess-get-words-from-vector-test
   passed   24/125  ess-help-aliases-test
   passed   25/125  ess-imenu-test
   passed   26/125  ess-inf-send-cat-some\.text-test
   passed   27/125  ess-inf-send-complex-input-test
   passed   28/125  ess-inf-send-fn-test
   passed   29/125  ess-inf-send-input-invisible-test
   passed   30/125  ess-inferior-force-test
   passed   31/125  ess-insert-assign-test
   passed   32/125  ess-jags-mode
   passed   33/125  ess-julia-mode
   passed   34/125  ess-load-file-test
   passed   35/125  ess-mode-inherits-prog-mode
   passed   36/125  ess-narrow-to-defun-or-para-test
   passed   37/125  ess-next-code-line-test
   passed   38/125  ess-r-beggining/end-of-defun-ignore-inner-fn-test
   passed   39/125  ess-r-beginning/defun-limits-test
   passed   40/125  ess-r-beginning/function-limits-test-1
   passed   41/125  ess-r-beginning/function-limits-test-2
   passed   42/125  ess-r-commands-test
   passed   43/125  ess-r-comment-dwim-test
   passed   44/125  ess-r-eval-linewise-visibly-test
   passed   45/125  ess-r-eval-ns-env-roxy-standard-test
   passed   46/125  ess-r-eval-ns-env-roxy-tracebug-test
   passed   47/125  ess-r-eval-rectangle-mark-mode-test
   passed   48/125  ess-r-eval-sink-freeze-test
   passed   49/125  ess-r-eval-visibility-eval-standard-filter-test
Test ess-r-eval-visibility-eval-test backtrace:
  signal(ert-test-failed (((should (string= actual expected)) :form (string= "> 1\n[1] 1\n> 2\n[1] 2\n> {\n+   3\n+   4\n+ }\n[1] 
  ert-fail(((should (string= actual expected)) :form (string= "> 1\n[1] 1\n> 2\n[1] 2\n> {\n+   3\n+   4\n+ }\n[1] 4\n> 5; 6\n[1] 
  (if (unwind-protect (setq value-1069 (apply fn-1067 args-1068)) (setq form-description-1071 (nconc (list (quote (should (string=
  (let (form-description-1071) (if (unwind-protect (setq value-1069 (apply fn-1067 args-1068)) (setq form-description-1071 (nconc 
  (let ((value-1069 (quote ert-form-evaluation-aborted-1070))) (let (form-description-1071) (if (unwind-protect (setq value-1069 (
  (let* ((fn-1067 (function string=)) (args-1068 (condition-case err (let ((signal-hook-function (function ert--should-signal-hook
  (closure (t) (actual expected) (let* ((fn-1067 (function string=)) (args-1068 (condition-case err (let ((signal-hook-function ..
  funcall((closure (t) (actual expected) (let* ((fn-1067 (function string=)) (args-1068 (condition-case err (let (...) (list actua
  (save-current-buffer (set-buffer inf-buf) (save-excursion (let ((times 0)) (while (and (< times 10) (not (re-search-forward "> \
  (let* ((inf-buf etest-local-inferior-buffer) (inf-proc (get-buffer-process inf-buf))) (save-current-buffer (set-buffer inf-buf) 
  (unwind-protect (let* ((inf-buf etest-local-inferior-buffer) (inf-proc (get-buffer-process inf-buf))) (save-current-buffer (set-
  etest--flush-inferior-buffer((closure (t) (actual expected) (let* ((fn-1067 (function string=)) (args-1068 (condition-case err (
  (cond ((eq etest--key :inf-buffer) (setq etest-local-inferior-buffer (eval etest--value))) ((eq etest--key :cleanup) (setq etest
  (let ((etest--key (car-safe (prog1 body (setq body (cdr body))))) (etest--value (car-safe (prog1 body (setq body (cdr body))))))
  (while body (let ((etest--key (car-safe (prog1 body (setq body (cdr body))))) (etest--value (car-safe (prog1 body (setq body (cd
  (unwind-protect (while body (let ((etest--key (car-safe (prog1 body (setq body ...)))) (etest--value (car-safe (prog1 body (setq
  (let ((etest--msg-sentinel (etest--make-message-sentinel)) etest--cleanup) (unwind-protect (while body (let ((etest--key (car-sa
  (save-current-buffer (set-buffer etest--buf) (let ((etest--msg-sentinel (etest--make-message-sentinel)) etest--cleanup) (unwind-
  (unwind-protect (save-current-buffer (set-buffer etest--buf) (let ((etest--msg-sentinel (etest--make-message-sentinel)) etest--c
  (let ((etest--buf (etest--new-buffer (let (local) (while (eq (car body) :init) (car-safe (prog1 body ...)) (setq local (append l
  etest--run-test((:init ((mode . r) (eval ess-test-r-set-local-process (quote tracebug)) (ess-eval-deactivate-mark)) :case "\n¶1\
  (closure (t) nil (etest--run-test (quote (:init ((mode . r) (eval ess-test-r-set-local-process (quote tracebug)) (ess-eval-deact
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test :name ess-r-eval-visibility-eval-test :documentation "`ess-
  ert-run-test(#s(ert-test :name ess-r-eval-visibility-eval-test :documentation "`ess-eval-region' respects `ess-eval-visibly'.\nD
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test :name R-mode :documentation nil :body (closure (t) nil (let*
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #<bytecode 0x10028b491>) nil)
  ert-run-tests-batch(t)
  ert-run-tests-batch-and-exit(t)
  eval-buffer(#<buffer  *load*> nil "c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t)  ; Reading at buffer position 23
  load-with-code-conversion("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" "c:/Users/r2/Projects/github/ESS/_1025/test/run
  load("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t t)
  command-line-1(("-scriptload" "run-tests"))
  command-line()
  normal-top-level()
Test ess-r-eval-visibility-eval-test condition:
    (ert-test-failed
     ((should
       (string= actual expected))
      :form
      (string= "> 1
[1] 1
> 2
[1] 2
> {
+   3
+   4
+ }
[1] 4
> 5; 6
[1] 5
[1] 6
> " "1
[1] 1
> 2
[1] 2
> {
+   3
+   4
+ }
[1] 4
> 5; 6
[1] 5
[1] 6
> ")
      :value nil))
   FAILED   50/125  ess-r-eval-visibility-eval-test
   passed   51/125  ess-r-function-beg-end-test
   passed   52/125  ess-r-goto-beginning/end-of-function-or-para-test
  skipped   53/125  ess-r-help-usage-objects-test
   passed   54/125  ess-r-inherits-prog-mode-test
   passed   55/125  ess-r-makevars-mode-test
   passed   56/125  ess-r-namespaced-eval-no-sourced-message-test
   passed   57/125  ess-r-namespaced-eval-no-srcref-in-errors-test
   passed   58/125  ess-r-no-user-style-overwrite-test
   passed   59/125  ess-r-package-auto-activation-in-shell-test
   passed   60/125  ess-r-package-auto-activation-test
  skipped   61/125  ess-r-package-auto-no-activation-in-shell-test
   passed   62/125  ess-r-package-package-info-test
Test ess-r-package-vars-test backtrace:
  process-send-string(nil "\n")
  ess-interrupt()
  inferior-ess-reload()
  (let ((pkg-dir (file-truename (cdr (ess-r-package-project)))) ess-ask-for-ess-directory) (ess-set-working-directory (expand-file
  (progn (let ((r-setwd-cmd (cdr (assq (quote ess-setwd-command) ess-r-customize-alist))) (r-getwd-cmd (cdr (assq (quote ess-getwd
  (unwind-protect (progn (let ((r-setwd-cmd (cdr (assq (quote ess-setwd-command) ess-r-customize-alist))) (r-getwd-cmd (cdr (assq 
  (let* ((inf-buf (run-ess-test-r-vanilla)) (inf-proc (get-buffer-process inf-buf)) (ess-local-process-name (process-name inf-proc
  (progn (set-buffer (if *file* (find-file-noselect *file*) (generate-new-buffer " *with-r-file-temp*"))) (c-mode) (let* ((inf-buf
  (unwind-protect (progn (set-buffer (if *file* (find-file-noselect *file*) (generate-new-buffer " *with-r-file-temp*"))) (c-mode)
  (let ((wconfig (current-window-configuration))) (unwind-protect (progn (set-buffer (if *file* (find-file-noselect *file*) (gener
  (let ((inhibit-message ess-inhibit-message-in-tests) (*file* "dummy-pkg/src/test.c")) (let ((wconfig (current-window-configurati
  (closure (t) nil (let ((inhibit-message ess-inhibit-message-in-tests) (*file* "dummy-pkg/src/test.c")) (let ((wconfig (current-w
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test :name ess-r-package-vars-test :documentation nil :body (clo
  ert-run-test(#s(ert-test :name ess-r-package-vars-test :documentation nil :body (closure (t) nil (let ((inhibit-message ess-inhi
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test :name R-mode :documentation nil :body (closure (t) nil (let*
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #<bytecode 0x10028b491>) nil)
  ert-run-tests-batch(t)
  ert-run-tests-batch-and-exit(t)
  eval-buffer(#<buffer  *load*> nil "c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t)  ; Reading at buffer position 23
  load-with-code-conversion("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" "c:/Users/r2/Projects/github/ESS/_1025/test/run
  load("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t t)
  command-line-1(("-scriptload" "run-tests"))
  command-line()
  normal-top-level()
Test ess-r-package-vars-test condition:
    (error "Buffer test.c has no process")
   FAILED   63/125  ess-r-package-vars-test
Test ess-r-pager-test backtrace:
  signal(ert-test-failed (((should (string= (car (ess-get-words-from-vector "getOption('pager')\n")) "cat")) :form (string= "conso
  ert-fail(((should (string= (car (ess-get-words-from-vector "getOption('pager')\n")) "cat")) :form (string= "console" "cat") :val
  (if (unwind-protect (setq value-1354 (apply fn-1352 args-1353)) (setq form-description-1356 (nconc (list (quote (should (string=
  (unless (unwind-protect (setq value-1354 (apply fn-1352 args-1353)) (setq form-description-1356 (nconc (list (quote (should (str
  (let (form-description-1356) (unless (unwind-protect (setq value-1354 (apply fn-1352 args-1353)) (setq form-description-1356 (nc
  (let ((value-1354 (quote ert-form-evaluation-aborted-1355))) (let (form-description-1356) (unless (unwind-protect (setq value-13
  (let* ((fn-1352 (function string=)) (args-1353 (condition-case err (let ((signal-hook-function (function ert--should-signal-hook
  (should (string= (car (ess-get-words-from-vector "getOption('pager')\n")) "cat"))
  eval((should (string= (car (ess-get-words-from-vector "getOption('pager')\n")) "cat")))
  (let ((inhibit-message t)) (eval x))
  (cond ((equal x (quote (kbd "C-u"))) (setq current-prefix-arg (list 4))) ((stringp x) (if (string= x "C-u") (setq current-prefix
  (closure ((cursor . 1) (--dolist-tail-- 1) (current-cpa) (current-lc . ess-eval-region) (cursors-end) (cursors-start 1) (reset-s
  mapc((closure ((cursor . 1) (--dolist-tail-- 1) (current-cpa) (current-lc . ess-eval-region) (cursors-end) (cursors-start 1) (re
  (let ((cursor (car --dolist-tail--))) (goto-char cursor) (setq last-command current-lc) (setq current-prefix-arg current-cpa) (m
  (while --dolist-tail-- (let ((cursor (car --dolist-tail--))) (goto-char cursor) (setq last-command current-lc) (setq current-pre
  (let ((--dolist-tail-- cursors-start)) (while --dolist-tail-- (let ((cursor (car --dolist-tail--))) (goto-char cursor) (setq las
  (let ((current-lc last-command) (current-cpa current-prefix-arg)) (let ((--dolist-tail-- cursors-start)) (while --dolist-tail-- 
  (let (cursors-start cursors-end) (while (search-backward "¶" nil t) (delete-char 1) (let ((marker (point-marker))) (set-marker-i
  (save-current-buffer (set-buffer buf) (goto-char (point-min)) (if (search-forward "×" nil t) (progn (backward-delete-char 1) (se
  etest-run(#<killed buffer> ((should (string= (car (ess-get-words-from-vector "getOption('pager')\n")) "cat"))))
  (cond ((eq etest--key :inf-buffer) (setq etest-local-inferior-buffer (eval etest--value))) ((eq etest--key :cleanup) (setq etest
  (let ((etest--key (car-safe (prog1 body (setq body (cdr body))))) (etest--value (car-safe (prog1 body (setq body (cdr body))))))
  (while body (let ((etest--key (car-safe (prog1 body (setq body (cdr body))))) (etest--value (car-safe (prog1 body (setq body (cd
  (unwind-protect (while body (let ((etest--key (car-safe (prog1 body (setq body ...)))) (etest--value (car-safe (prog1 body (setq
  (let ((etest--msg-sentinel (etest--make-message-sentinel)) etest--cleanup) (unwind-protect (while body (let ((etest--key (car-sa
  (save-current-buffer (set-buffer etest--buf) (let ((etest--msg-sentinel (etest--make-message-sentinel)) etest--cleanup) (unwind-
  (unwind-protect (save-current-buffer (set-buffer etest--buf) (let ((etest--msg-sentinel (etest--make-message-sentinel)) etest--c
  (let ((etest--buf (etest--new-buffer (let (local) (while (eq (car body) :init) (car-safe (prog1 body ...)) (setq local (append l
  etest--run-test((:init ((mode . r) (eval ess-test-r-set-local-process)) :eval ((should (string= (car (ess-get-words-from-vector 
  (closure (t) nil (etest--run-test (quote (:init ((mode . r) (eval ess-test-r-set-local-process)) :eval ((should (string= (car ..
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test :name ess-r-pager-test :documentation nil :body (closure (t
  ert-run-test(#s(ert-test :name ess-r-pager-test :documentation nil :body (closure (t) nil (etest--run-test (quote (:init ((mode 
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test :name R-mode :documentation nil :body (closure (t) nil (let*
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #<bytecode 0x10028b491>) nil)
  ert-run-tests-batch(t)
  ert-run-tests-batch-and-exit(t)
  eval-buffer(#<buffer  *load*> nil "c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t)  ; Reading at buffer position 23
  load-with-code-conversion("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" "c:/Users/r2/Projects/github/ESS/_1025/test/run
  load("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t t)
  command-line-1(("-scriptload" "run-tests"))
  command-line()
  normal-top-level()
Test ess-r-pager-test condition:
    (ert-test-failed
     ((should
       (string=
	(car ...)
	"cat"))
      :form
      (string= "console" "cat")
      :value nil))
   FAILED   64/125  ess-r-pager-test
   passed   65/125  ess-r-run-presend-hooks-test
   passed   66/125  ess-r-send-double-quoted-strings-test
   passed   67/125  ess-r-send-single-quoted-strings-test
   passed   68/125  ess-r-transcript-motions-test
   passed   69/125  ess-rd-eval-linewise-visibly-test
   passed   70/125  ess-rd-eval-ns-env-test
   passed   71/125  ess-roxy-cpp-test
   passed   72/125  ess-roxy-get-function-args-test
   passed   73/125  ess-roxy-preview-Rd-test
   passed   74/125  ess-roxy-update-entry-test
   passed   75/125  ess-run-presend-hooks-test
Test ess-set-working-directory-test backtrace:
  signal(ert-test-failed ("Expected: \nsetwd('/')\n> [1] \"/\"" "Result: \nsetwd('/')\n> [1] \"C:/\""))
  (if (string= output expected) output (signal (quote ert-test-failed) (list (concat "Expected: \n" expected) (concat "Result: \n"
  (let ((output (progn (ess-wait-for-process *proc*) (ess-eval-buffer nil) (ess-wait-for-process *proc*) (save-current-buffer (set
  (progn (let ((output (progn (ess-wait-for-process *proc*) (ess-eval-buffer nil) (ess-wait-for-process *proc*) (save-current-buff
  (setq value-436 (progn (let ((output (progn (ess-wait-for-process *proc*) (ess-eval-buffer nil) (ess-wait-for-process *proc*) (s
  (unwind-protect (setq value-436 (progn (let ((output (progn (ess-wait-for-process *proc*) (ess-eval-buffer nil) (ess-wait-for-pr
  (if (unwind-protect (setq value-436 (progn (let ((output (progn ... ... ... ...)) (expected (eval "setwd('/')\n> [1] \"/\""))) (
  (let (form-description-437) (if (unwind-protect (setq value-436 (progn (let ((output ...) (expected ...)) (if (string= output ex
  (let ((value-436 (gensym "ert-form-evaluation-aborted-"))) (let (form-description-437) (if (unwind-protect (setq value-436 (prog
  (progn (ess-set-working-directory "/") (ess-eval-linewise "getwd()" (quote invisible)) (let ((value-436 (gensym "ert-form-evalua
  (prog1 (progn (ess-set-working-directory "/") (ess-eval-linewise "getwd()" (quote invisible)) (let ((value-436 (gensym "ert-form
  (progn (setq ess-r-tests-current-output-buffer *inf-buf*) (let ((inhibit-read-only t)) (save-current-buffer (set-buffer ess-r-te
  (unwind-protect (progn (setq ess-r-tests-current-output-buffer *inf-buf*) (let ((inhibit-read-only t)) (save-current-buffer (set
  (unwind-protect (unwind-protect (progn (setq ess-r-tests-current-output-buffer *inf-buf*) (let ((inhibit-read-only t)) (save-cur
  (let* ((*proc* (get-buffer-process (run-ess-test-r-vanilla))) (ess-local-process-name (process-name *proc*)) (*inf-buf* (process
  (progn (switch-to-buffer r-file-buffer) (R-mode) (let* ((*proc* (get-buffer-process (run-ess-test-r-vanilla))) (ess-local-proces
  (unwind-protect (progn (switch-to-buffer r-file-buffer) (R-mode) (let* ((*proc* (get-buffer-process (run-ess-test-r-vanilla))) (
  (let ((wconfig (current-window-configuration))) (unwind-protect (progn (switch-to-buffer r-file-buffer) (R-mode) (let* ((*proc* 
  (let* ((inhibit-message ess-inhibit-message-in-tests) (buffer-or-file nil) (r-file-buffer (cond ((bufferp buffer-or-file) buffer
  (closure (t) nil (let* ((inhibit-message ess-inhibit-message-in-tests) (buffer-or-file nil) (r-file-buffer (cond ((bufferp buffe
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test :name ess-set-working-directory-test :documentation nil :bo
  ert-run-test(#s(ert-test :name ess-set-working-directory-test :documentation nil :body (closure (t) nil (let* ((inhibit-message 
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test :name R-mode :documentation nil :body (closure (t) nil (let*
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #<bytecode 0x10028b491>) nil)
  ert-run-tests-batch(t)
  ert-run-tests-batch-and-exit(t)
  eval-buffer(#<buffer  *load*> nil "c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t)  ; Reading at buffer position 23
  load-with-code-conversion("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" "c:/Users/r2/Projects/github/ESS/_1025/test/run
  load("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t t)
  command-line-1(("-scriptload" "run-tests"))
  command-line()
  normal-top-level()
Test ess-set-working-directory-test condition:
    (ert-test-failed "Expected: 
setwd('/')
> [1] \"/\"" "Result: 
setwd('/')
> [1] \"C:/\"")
   FAILED   76/125  ess-set-working-directory-test
   passed   77/125  ess-setwd-test
   passed   78/125  ess-skip-thing-test
   passed   79/125  ess-startup-default-directory-preserved-test
Test ess-startup-verbose-setwd-test backtrace:
  signal(ert-test-failed (((should (string-match "to quit R.\n\n> setwd(.*)$" (ess-r-tests-startup-output))) :form (string-match "
  ert-fail(((should (string-match "to quit R.\n\n> setwd(.*)$" (ess-r-tests-startup-output))) :form (string-match "to quit R.\n\n>
  (if (unwind-protect (setq value-41 (apply fn-39 args-40)) (setq form-description-43 (nconc (list (quote (should (string-match "t
  (let (form-description-43) (if (unwind-protect (setq value-41 (apply fn-39 args-40)) (setq form-description-43 (nconc (list (quo
  (let ((value-41 (quote ert-form-evaluation-aborted-42))) (let (form-description-43) (if (unwind-protect (setq value-41 (apply fn
  (let* ((fn-39 (function string-match)) (args-40 (condition-case err (let ((signal-hook-function (function ert--should-signal-hoo
  (closure (t) nil (let* ((fn-39 (function string-match)) (args-40 (condition-case err (let ((signal-hook-function ...)) (list "to
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test :name ess-startup-verbose-setwd-test :documentation nil :bo
  ert-run-test(#s(ert-test :name ess-startup-verbose-setwd-test :documentation nil :body (closure (t) nil (let* ((fn-39 (function 
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test :name R-mode :documentation nil :body (closure (t) nil (let*
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #<bytecode 0x10028b491>) nil)
  ert-run-tests-batch(t)
  ert-run-tests-batch-and-exit(t)
  eval-buffer(#<buffer  *load*> nil "c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t)  ; Reading at buffer position 23
  load-with-code-conversion("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" "c:/Users/r2/Projects/github/ESS/_1025/test/run
  load("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t t)
  command-line-1(("-scriptload" "run-tests"))
  command-line()
  normal-top-level()
Test ess-startup-verbose-setwd-test condition:
    (ert-test-failed
     ((should
       (string-match "to quit R.

> setwd(.*)$"
		     (ess-r-tests-startup-output)))
      :form
      (string-match "to quit R.

> setwd(.*)$"
		    #("
R version 4.0.3 (2020-10-10) -- \"Bunny-Wunnies Freak Out\"
Copyright (C) 2020 The R Foundation for Statistical Computing
Platform: x86_64-w64-mingw32/x64 (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

During startup - Warning message:
Setting LC_CTYPE=en_US.UTF-8 failed 
> setwd('~/Projects/github/ESS/_1025/test')
> " 0 702
(fontified nil)
702 773
(fontified nil)
773 775
(fontified nil rear-nonsticky t)
775 816
(field input front-sticky t font-lock-face comint-highlight-input fontified nil)
816 817
(font-lock-face comint-highlight-input fontified nil)
817 819
(fontified nil font-lock-face ... rear-nonsticky t)))
      :value nil))
   FAILED   80/125  ess-startup-verbose-setwd-test
   passed   81/125  ess-switch-to-inferior-or-script-buffer-test
   passed   82/125  ess-test-inferior-live-process-error
Test ess-test-inferior-local-start-args backtrace:
  signal(ert-test-failed (((should (equal (car inf-data) "R")) :form (equal "Rterm" "R") :value nil :explanation (arrays-of-differ
  ert-fail(((should (equal (car inf-data) "R")) :form (equal "Rterm" "R") :value nil :explanation (arrays-of-different-length 5 1 
  (if (unwind-protect (setq value-73 (apply fn-71 args-72)) (setq form-description-75 (nconc (list (quote (should (equal ... "R"))
  (let (form-description-75) (if (unwind-protect (setq value-73 (apply fn-71 args-72)) (setq form-description-75 (nconc (list (quo
  (let ((value-73 (quote ert-form-evaluation-aborted-74))) (let (form-description-75) (if (unwind-protect (setq value-73 (apply fn
  (let* ((fn-71 (function equal)) (args-72 (condition-case err (let ((signal-hook-function (function ert--should-signal-hook))) (l
  (let ((inf-data (buffer-local-value (quote inferior-ess--local-data) *inf-buf*))) (let* ((fn-71 (function equal)) (args-72 (cond
  (progn (let ((inf-data (buffer-local-value (quote inferior-ess--local-data) *inf-buf*))) (let* ((fn-71 (function equal)) (args-7
  (prog1 (progn (let ((inf-data (buffer-local-value (quote inferior-ess--local-data) *inf-buf*))) (let* ((fn-71 (function equal)) 
  (progn (setq ess-r-tests-current-output-buffer *inf-buf*) (let ((inhibit-read-only t)) (save-current-buffer (set-buffer ess-r-te
  (unwind-protect (progn (setq ess-r-tests-current-output-buffer *inf-buf*) (let ((inhibit-read-only t)) (save-current-buffer (set
  (unwind-protect (unwind-protect (progn (setq ess-r-tests-current-output-buffer *inf-buf*) (let ((inhibit-read-only t)) (save-cur
  (let* ((*proc* (get-buffer-process (run-ess-test-r-vanilla))) (ess-local-process-name (process-name *proc*)) (*inf-buf* (process
  (progn (switch-to-buffer r-file-buffer) (R-mode) (let* ((*proc* (get-buffer-process (run-ess-test-r-vanilla))) (ess-local-proces
  (unwind-protect (progn (switch-to-buffer r-file-buffer) (R-mode) (let* ((*proc* (get-buffer-process (run-ess-test-r-vanilla))) (
  (let ((wconfig (current-window-configuration))) (unwind-protect (progn (switch-to-buffer r-file-buffer) (R-mode) (let* ((*proc* 
  (let* ((inhibit-message ess-inhibit-message-in-tests) (buffer-or-file nil) (r-file-buffer (cond ((bufferp buffer-or-file) buffer
  (closure (t) nil (let* ((inhibit-message ess-inhibit-message-in-tests) (buffer-or-file nil) (r-file-buffer (cond ((bufferp buffe
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test :name ess-test-inferior-local-start-args :documentation nil
  ert-run-test(#s(ert-test :name ess-test-inferior-local-start-args :documentation nil :body (closure (t) nil (let* ((inhibit-mess
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test :name R-mode :documentation nil :body (closure (t) nil (let*
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #<bytecode 0x10028b491>) nil)
  ert-run-tests-batch(t)
  ert-run-tests-batch-and-exit(t)
  eval-buffer(#<buffer  *load*> nil "c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t)  ; Reading at buffer position 23
  load-with-code-conversion("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" "c:/Users/r2/Projects/github/ESS/_1025/test/run
  load("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t t)
  command-line-1(("-scriptload" "run-tests"))
  command-line()
  normal-top-level()
Test ess-test-inferior-local-start-args condition:
    (ert-test-failed
     ((should
       (equal
	(car inf-data)
	"R"))
      :form
      (equal "Rterm" "R")
      :value nil :explanation
      (arrays-of-different-length 5 1 "Rterm" "R" first-mismatch-at 1)))
   FAILED   83/125  ess-test-inferior-local-start-args
Test ess-test-inferior-reload-start-data backtrace:
  signal(ert-test-failed (((should (equal (cdr inf-data) "--no-readline  --vanilla")) :form (equal "--ess  --vanilla" "--no-readli
  ert-fail(((should (equal (cdr inf-data) "--no-readline  --vanilla")) :form (equal "--ess  --vanilla" "--no-readline  --vanilla")
  (if (unwind-protect (setq value-88 (apply fn-86 args-87)) (setq form-description-90 (nconc (list (quote (should (equal ... "--no
  (let (form-description-90) (if (unwind-protect (setq value-88 (apply fn-86 args-87)) (setq form-description-90 (nconc (list (quo
  (let ((value-88 (quote ert-form-evaluation-aborted-89))) (let (form-description-90) (if (unwind-protect (setq value-88 (apply fn
  (let* ((fn-86 (function equal)) (args-87 (condition-case err (let ((signal-hook-function (function ert--should-signal-hook))) (l
  (let ((inf-data (buffer-local-value (quote inferior-ess--local-data) inf-buf))) (let* ((fn-81 (function equal)) (args-82 (condit
  (progn (let ((inf-data (buffer-local-value (quote inferior-ess--local-data) inf-buf))) (let* ((fn-81 (function equal)) (args-82 
  (unwind-protect (progn (let ((inf-data (buffer-local-value (quote inferior-ess--local-data) inf-buf))) (let* ((fn-81 (function e
  (let* ((r-path (executable-find "R")) (inferior-ess-r-program r-path) (inf-buf (run-ess-test-r-vanilla))) (unwind-protect (progn
  (closure (t) nil (let* ((r-path (executable-find "R")) (inferior-ess-r-program r-path) (inf-buf (run-ess-test-r-vanilla))) (unwi
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test :name ess-test-inferior-reload-start-data :documentation ni
  ert-run-test(#s(ert-test :name ess-test-inferior-reload-start-data :documentation nil :body (closure (t) nil (let* ((r-path (exe
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test :name R-mode :documentation nil :body (closure (t) nil (let*
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #<bytecode 0x10028b491>) nil)
  ert-run-tests-batch(t)
  ert-run-tests-batch-and-exit(t)
  eval-buffer(#<buffer  *load*> nil "c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t)  ; Reading at buffer position 23
  load-with-code-conversion("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" "c:/Users/r2/Projects/github/ESS/_1025/test/run
  load("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t t)
  command-line-1(("-scriptload" "run-tests"))
  command-line()
  normal-top-level()
Test ess-test-inferior-reload-start-data condition:
    (ert-test-failed
     ((should
       (equal
	(cdr inf-data)
	"--no-readline  --vanilla"))
      :form
      (equal "--ess  --vanilla" "--no-readline  --vanilla")
      :value nil :explanation
      (arrays-of-different-length 16 24 "--ess  --vanilla" "--no-readline  --vanilla" first-mismatch-at 2)))
   FAILED   84/125  ess-test-inferior-reload-start-data
   passed   85/125  ess-test-r-comint-input-ring-file-name
Test ess-test-r-help-mode backtrace:
  signal(user-error ("No more buttons"))
  user-error("No more buttons")
  forward-button(2)
  (let ((ess-pop-to-buffer t)) (ess-display-help-on-object "plot") (let* ((fn-872 (function equal)) (args-873 (condition-case err 
  (progn (let ((ess-pop-to-buffer t)) (ess-display-help-on-object "plot") (let* ((fn-872 (function equal)) (args-873 (condition-ca
  (prog1 (progn (let ((ess-pop-to-buffer t)) (ess-display-help-on-object "plot") (let* ((fn-872 (function equal)) (args-873 (condi
  (progn (setq ess-r-tests-current-output-buffer *inf-buf*) (let ((inhibit-read-only t)) (save-current-buffer (set-buffer ess-r-te
  (unwind-protect (progn (setq ess-r-tests-current-output-buffer *inf-buf*) (let ((inhibit-read-only t)) (save-current-buffer (set
  (unwind-protect (unwind-protect (progn (setq ess-r-tests-current-output-buffer *inf-buf*) (let ((inhibit-read-only t)) (save-cur
  (let* ((*proc* (get-buffer-process (run-ess-test-r-vanilla))) (ess-local-process-name (process-name *proc*)) (*inf-buf* (process
  (progn (switch-to-buffer r-file-buffer) (R-mode) (let* ((*proc* (get-buffer-process (run-ess-test-r-vanilla))) (ess-local-proces
  (unwind-protect (progn (switch-to-buffer r-file-buffer) (R-mode) (let* ((*proc* (get-buffer-process (run-ess-test-r-vanilla))) (
  (let ((wconfig (current-window-configuration))) (unwind-protect (progn (switch-to-buffer r-file-buffer) (R-mode) (let* ((*proc* 
  (let* ((inhibit-message ess-inhibit-message-in-tests) (buffer-or-file nil) (r-file-buffer (cond ((bufferp buffer-or-file) buffer
  (closure (t) nil (let* ((inhibit-message ess-inhibit-message-in-tests) (buffer-or-file nil) (r-file-buffer (cond ((bufferp buffe
  ert--run-test-internal(#s(ert--test-execution-info :test #s(ert-test :name ess-test-r-help-mode :documentation nil :body (closur
  ert-run-test(#s(ert-test :name ess-test-r-help-mode :documentation nil :body (closure (t) nil (let* ((inhibit-message ess-inhibi
  ert-run-or-rerun-test(#s(ert--stats :selector t :tests [#s(ert-test :name R-mode :documentation nil :body (closure (t) nil (let*
  ert-run-tests(t #f(compiled-function (event-type &rest event-args) #<bytecode 0x10028b491>) nil)
  ert-run-tests-batch(t)
  ert-run-tests-batch-and-exit(t)
  eval-buffer(#<buffer  *load*> nil "c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t)  ; Reading at buffer position 23
  load-with-code-conversion("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" "c:/Users/r2/Projects/github/ESS/_1025/test/run
  load("c:/Users/r2/Projects/github/ESS/_1025/test/run-tests" nil t t)
  command-line-1(("-scriptload" "run-tests"))
  command-line()
  normal-top-level()
Test ess-test-r-help-mode condition:
    (user-error "No more buttons")
   FAILED   86/125  ess-test-r-help-mode
  skipped   87/125  ess-test-r-index-mode
   passed   88/125  ess-test-roxy-font-lock
   passed   89/125  ess-test-roxy-prefix-strip
   passed   90/125  ess-transcript-comint-prompt-test
   passed   91/125  etest-cleanup-test
   passed   92/125  etest-climb-deftest-test
   passed   93/125  etest-inferior-buffer-test
   passed   94/125  etest-keep-state-test
   passed   95/125  etest-local-vars-test
   passed   96/125  etest-mark-test
   passed   97/125  etest-messages-test
   passed   98/125  etest-multiple-results-test
   passed   99/125  etest-skip-comments-test
   passed  100/125  etest-test-result-test
   passed  101/125  etest-unalias-prefix-key
   passed  102/125  etest-update-result-not-on-bol-test
   passed  103/125  etest-update-test
   passed  104/125  etest-wrap-test-keyword-test
   passed  105/125  inferior-ess-inherits-from-comint-test
   passed  106/125  inferior-ess-r-fontification-test
  skipped  107/125  runner-R-3\.2\.1-buffer-name-test
  skipped  108/125  runner-R-3\.2\.1-defined-test
   passed  109/125  sas-mode
   passed  110/125  test-elt
   passed  111/125  test-ess-R-indentation-C++
   passed  112/125  test-ess-R-indentation-RRR
   passed  113/125  test-ess-R-indentation-RRR+
   passed  114/125  test-ess-R-indentation-RStudio-
   passed  115/125  test-ess-R-indentation-misc1
   passed  116/125  test-ess-r-fontification
   passed  117/125  test-ess-r-misc
   passed  118/125  test-ess-r-syntax
   passed  119/125  test-ess-r-tokens
   passed  120/125  test-ess-roxy-literate
   passed  121/125  test-org-ob-R-data-frame-test
   passed  122/125  test-org-ob-R-output-test
   passed  123/125  test-org-ob-R-session-output-test
   passed  124/125  test-org-ob-R-session-value-test
   passed  125/125  test-org-ob-R-value-test

Ran 125 tests, 111 results as expected, 9 unexpected, 5 skipped (2020-12-30 11:45:55-0500)

9 unexpected results:
   FAILED  ess--command-browser-timeout-test
   FAILED  ess-r-eval-visibility-eval-test
   FAILED  ess-r-package-vars-test
   FAILED  ess-r-pager-test
   FAILED  ess-set-working-directory-test
   FAILED  ess-startup-verbose-setwd-test
   FAILED  ess-test-inferior-local-start-args
   FAILED  ess-test-inferior-reload-start-data
   FAILED  ess-test-r-help-mode

5 skipped results:
  SKIPPED  ess-r-help-usage-objects-test
  SKIPPED  ess-r-package-auto-no-activation-in-shell-test
  SKIPPED  ess-test-r-index-mode
  SKIPPED  runner-R-3\.2\.1-buffer-name-test
  SKIPPED  runner-R-3\.2\.1-defined-test

make: *** [Makefile:12: lisp] Error 1
```

</details>

Reply:
I agree that usage sections would be enough for these tests.

Thanks for the make output, it's not too bad actually!

Reply:
Actually, I was thinking ... instead of storing the already-converted text, it will be more resilient to store the `.Rd` format and have R parse it during testing. This should provide assurance that if R ever changes the layout, the tests will likely start to fail, indicating the pattern needs to change anyway.

`capture.output(tools::Rd2txt(filepath))` is good for this. (I think ESS may be doing this already?)

Are you good with using `.Rd`-format files instead of `.txt`?

---

Side note: on windows at least, I had to preface it with `tools::Rd2txt_options(underline_titles = FALSE)` to preclude overstriking (which defeats the regex patterns); without it, `"Usage:"` becomes `"_\bU_\bs_\ba_\bg_\be:"` (and fails to match). We can update the matching pattern from the first below to the second:

```elisp
(re-search-forward "^Usage:" nil t)
(re-search-forward "^\\(_\b\\)?U\\(_\b\\)?s\\(_\b\\)?a\\(_\b\\)?g\\(_\b\\)?e:" nil t)
```

While that works, it is *ugly*, and I'm guessing ultimately unnecessary.

Reply:
(Forget it, I'll mimick `ess-r-help-usage-objects-test` in a singleton unit-test.)

Reply:
Well, not for now.  We really are about fixing bugs / issues to get a stable ESS we can finally "formally release"
(together with `polymode-R` etc in an  `ESS+` bundle)   ... hopefully within a few weeks time.

After that, I believe ESS should consider using R to parse the *.Rd files... This would most importantly also help to produce correct translation of `\link{}` entries .. which we currently do from the *.txt files and hence are *wrong* about 1% of times.